src/Algebra/Graph/LabelledAdjacencyMap.hs

```diff
@@ -0,0 +1,780 @@
+
+-- | The /tree graph/ constructed from a given 'Tree' data structure.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- tree (Node x [])                                         == 'vertex' x
+-- tree (Node x [Node y [Node z []]])                       == 'path' [x,y,z]
+-- tree (Node x [Node y [], Node z []])                     == 'star' x [y,z]
+-- tree (Node 1 [Node 2 [], Node 3 [Node 4 [], Node 5 []]]) == 'edges' [(1,2), (1,3), (3,4), (3,5)]
+-- @
+tree :: (Ord a, Dioid e) => Tree a -> LabelledAdjacencyMap a e
+tree (Node x []) = vertex x
+tree (Node x f ) = star x (map rootLabel f)
+    `overlay` forest (filter (not . null . subForest) f)
+
+-- | The /forest graph/ constructed from a given 'Forest' data structure.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- forest []                                                  == 'empty'
+-- forest [x]                                                 == 'tree' x
+-- forest [Node 1 [Node 2 [], Node 3 []], Node 4 [Node 5 []]] == 'edges' [(1,2), (1,3), (4,5)]
+-- forest                                                     == 'overlays' . map 'tree'
+-- @
+forest :: (Ord a, Dioid e) => Forest a -> LabelledAdjacencyMap a e
+forest = overlays . map tree
+
+-- | Remove a vertex from a given graph.
+-- Complexity: /O(n*log(n))/ time.
+--
+-- @
+-- removeVertex x ('vertex' x)       == 'empty'
+-- removeVertex 1 ('vertex' 2)       == 'vertex' 2
+-- removeVertex x ('edge' x x)       == 'empty'
+-- removeVertex 1 ('edge' 1 2)       == 'vertex' 2
+-- removeVertex x . removeVertex x == removeVertex x
+-- @
+removeVertex
+        :: Ord a => a -> LabelledAdjacencyMap a e -> LabelledAdjacencyMap a e
+removeVertex x =
+        LAM . Map.map (Map.delete x) . Map.delete x . labelledAdjacencyMap
+
+-- | Remove an edge from a given graph.
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- removeEdge x y ('edge' x y)       == 'vertices' [x, y]
+-- removeEdge x y . removeEdge x y == removeEdge x y
+-- removeEdge x y . 'removeVertex' x == 'removeVertex' x
+-- removeEdge 1 1 (1 * 1 * 2 * 2)  == 1 * 2 * 2
+-- removeEdge 1 2 (1 * 1 * 2 * 2)  == 1 * 1 + 2 * 2
+-- @
+removeEdge
+        :: Ord a
+        => a
+        -> a
+        -> LabelledAdjacencyMap a e
+        -> LabelledAdjacencyMap a e
+removeEdge x y = LAM . Map.adjust (Map.delete y) x . labelledAdjacencyMap
+
+-- | The function @'replaceVertex' x y@ replaces vertex @x@ with vertex @y@ in a
+-- given 'AdjacencyMap'. If @y@ already exists, @x@ and @y@ will be merged.
+-- Complexity: /O((n + m) * log(n))/ time.
+--
+-- @
+-- replaceVertex x x            == id
+-- replaceVertex x y ('vertex' x) == 'vertex' y
+-- replaceVertex x y            == 'mergeVertices' (== x) y
+-- @
+replaceVertex
+        :: Ord a
+        => a
+        -> a
+        -> LabelledAdjacencyMap a e
+        -> LabelledAdjacencyMap a e
+replaceVertex u v = gmap $ \w -> if w == u then v else w
+-- TODO: replaceVertexWithLabel?
```

Comment:
Not sure what would `replaceVertexWithLabel` do?

Reply:
These are probably not general enough warrant the addition.   My thought was that if you replace a vertex, you may want to do something specific to its labels as well.